packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,318 @@
+    options?: RequestOptions,
+  ): Promise<z.infer<T>> {
+    return new Promise((resolve, reject) => {
+      if (!this._transport) {
+        reject(new Error('Not connected'));
+        return;
+      }
+
+      options?.signal?.throwIfAborted();
+
+      const messageId = this._requestMessageId++;
+      const jsonrpcRequest: JSONRPCRequest = {
+        ...request,
+        jsonrpc: '2.0',
+        id: messageId,
+      };
+
+      const cancel = (reason: unknown) => {};
+
+      this._responseHandlers.set(messageId, response => {
+        if (options?.signal?.aborted) return;
+
+        if (response instanceof Error) {
+          return reject(response);
+        }
+
+        try {
+          const result = resultSchema.parse(response.result);
+          resolve(result);
+        } catch (error) {
+          reject(error);
+        }
+      });
+
+      options?.signal?.addEventListener('abort', () => {
+        cancel(options?.signal?.reason);
+      });
+
+      const timeout = options?.timeout ?? this.requestTimeoutMs;
+      const timeoutHandler = () =>
+        cancel(
+          new AISDKError({
+            name: 'McpClientTimeoutError',
+            message: 'Request timed out',
+          }),
+        );
+
+      this._setupTimeout(
+        messageId,
+        timeout,
+        options?.maxTotalTimeout,
+        timeoutHandler,
+      );
+
+      this._transport.send(jsonrpcRequest).catch(error => {
+        this._cleanupTimeout(messageId);
+        reject(error);
+      });
+    });
+  }
+
+  async listTools(
+    params?: ListToolsRequest['params'],
+    options?: RequestOptions,
+  ): Promise<ListToolsResult> {
+    if (!this._serverCapabilities?.tools) {
+      throw new AISDKError({
+        name: 'McpClientError',
+        message: `Server does not support tools (required for tools/list)`,
+      });
+    }
+
+    return this.request(
+      { method: 'tools/list', params },
+      ListToolsResultSchema,
+      options,
+    );
+  }
+
+  async callTool(
+    params: CallToolRequest['params'],
+    resultSchema:
+      | typeof CallToolResultSchema
+      | typeof CompatibilityCallToolResultSchema = CallToolResultSchema,
+    options?: RequestOptions,
+  ): Promise<CallToolResult | CompatibilityCallToolResult> {
+    if (!this._serverCapabilities?.tools) {
+      throw new AISDKError({
+        name: 'McpClientError',
+        message: `Server does not support tools (required for tools/call)`,
+      });
+    }
+
+    return this.request(
+      { method: 'tools/call', params },
+      resultSchema,
+      options,
+    );
+  }
+
+  async notification(notification: Notification): Promise<void> {
+    if (!this._transport) {
+      throw new AISDKError({
+        name: 'McpClientError',
+        message: 'Not connected',
+      });
+    }
+
+    const jsonrpcNotification: JSONRPCNotification = {
+      ...notification,
+      jsonrpc: '2.0',
+    };
+
+    await this._transport.send(jsonrpcNotification);
+  }
```

Comment:
do we need this?

Reply:
yes! this is needed to finalize the handshake w/ the server after we send the `initialize` request:

```
await this.notification({
        method: 'notifications/initialized',
});
```